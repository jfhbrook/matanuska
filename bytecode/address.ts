export type Address = number;
export type Pointer = Address | null;
export type Offset = Address;
export type ByteAddress = [number, number];

export function addrToBytes(addr: Address): ByteAddress {
  return [(addr >> 8) & 0xff, addr & 0xff];
}

export function bytesToAddr([first, second]: ByteAddress): Address {
  return (first << 8) | second;
}
