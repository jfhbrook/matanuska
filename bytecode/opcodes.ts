export enum OpCode {
  // Constants and values
  Constant,
  Nil,
  True,
  False,

  // !
  Pop,

  // Operators
  //
  // TODO: BASIC is typed - we can check the types in the compiler and use
  // typed op codes. However, this requires tracking the state of the stack
  // in the compiler. I'm keeping the opcodes untyped for now, just to get
  // something working, but would like to switch to typed opcodes later.

  // Comparison operators
  Eq,
  Gt,
  Ge,
  Lt,
  Le,
  Ne,

  // Operators
  Not,
  Add,
  Sub,
  Mul,
  Div,
  Neg,

  // Commands
  Print,

  // Jumping
  Jump,
  JumpIfFalse,
  Loop,

  Return,
}
