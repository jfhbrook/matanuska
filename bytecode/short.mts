import { Byte } from './byte.mjs';

export type Short = number;
export type Bytes = [Byte, Byte];

export function shortToBytes(short: Short): Bytes {
  return [(short >> 8) & 0xff, short & 0xff];
}

export function bytesToShort([first, second]: Bytes): Short {
  return (first << 8) | second;
}
