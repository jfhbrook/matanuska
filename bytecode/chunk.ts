import { Value } from '../value';

import { Byte } from './byte';
import { OpCode } from './opcodes';

export class Chunk {
  public filename: string = '<unknown>';
  public constants: Value[] = [];
  // TODO: In a C implementation, this would be an array of bytes. In clox,
  // the values are either opcodes, indexes to constants, or 16 byte offsets.
  //
  // In theory, I should be able to use an UInt8Array and do the same bit
  // shifting shenanigans as clox to handle 16 bit values. Using them is a
  // little challenging, though, and will require more research.
  //
  // For now, I'm using an Array to prototype.
  public code: Array<OpCode | Byte> = [];

  // The index is the same as the bytecode's. The value is the lineNo. Runtime
  // error reporting doesn't include the source line, just the line number
  // (and perhaps the name of the function).
  public lines: number[] = [];

  // Write an opcode, including its corresponding line number.
  public writeOp(op: OpCode, lineNo: number): void {
    this.code.push(op);
    this.lines.push(lineNo);
  }

  // Returns the index of the constant. I believe clox uses this number to
  // immediately write the ADD_CONSTANT op.
  public addConstant(value: Value): number {
    this.constants.push(value);
    return this.constants.length - 1;
  }
}
