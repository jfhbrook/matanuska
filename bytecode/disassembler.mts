import table from 'text-table';

import { formatter } from '../format.mjs';
import { bytesToShort } from './short.mjs';
import { OpCode } from './opcodes.mjs';
import { Chunk } from './chunk.mjs';

type LineNo = string;
type Offset = string;
type Code = string;
type Row = [LineNo, Offset, Code, string];

function chunkName(chunk: Chunk): string {
  return chunk.filename;
}

function toHex(value: number): string {
  return `0x${value.toString(16)}`;
}

export function disassemble(chunk: Chunk): string {
  const data: Row[] = [];
  const header = `=== Disassembly of ${chunkName(chunk)}: ===`;

  let offset = 0;
  let row: Row;

  while (offset < chunk.code.length) {
    [offset, row] = _disassembleInstruction(chunk, offset);
    data.push(row);
  }

  return `${header}\n${table(data)}`;
}

const FORWARD = 1;
// const BACKWARD = -1;

export function disassembleInstruction(chunk: Chunk, offset: number): string {
  const row = _disassembleInstruction(chunk, offset)[1];
  return table([row]);
}

function _disassembleInstruction(chunk: Chunk, offset: number): [number, Row] {
  let lineNo = 0;
  let code: number = 0xff;
  let row: Row;

  function advance(amount: number = 1): number {
    offset += amount;
    code = chunk.code[offset - 1];
    lineNo = chunk.lines[offset - 1];
    return code;
  }

  function peek(n: number = 0): number {
    return chunk.code[offset + n - 1];
  }

  function simple(code: string): Row {
    return [String(lineNo), String(offset), code, ''];
  }

  function constant(code: string): Row {
    return [
      String(lineNo),
      String(offset),
      code,
      formatter.format(chunk.constants[advance()]),
    ];
  }

  function jump(code: string, sign: number): Row {
    const jump = bytesToShort([peek(1), peek(2)]);

    const row: Row = [
      String(lineNo),
      String(offset),
      code,
      `${offset} -> ${offset + 3 + sign * jump}`,
    ];

    advance(2);

    return row;
  }

  switch (advance()) {
    case OpCode.Constant:
      row = constant('CONSTANT');
      break;
    case OpCode.Nil:
      row = simple('NIL');
      break;
    case OpCode.True:
      row = simple('TRUE');
      break;
    case OpCode.False:
      row = simple('FALSE');
      break;
    case OpCode.Pop:
      row = simple('POP');
      break;
    case OpCode.GetGlobal:
      row = constant('GET_GLOBAL');
      break;
    case OpCode.DefineGlobal:
      row = constant('DEFINE_GLOBAL');
      break;
    case OpCode.SetGlobal:
      row = constant('SET_GLOBAL');
      break;
    case OpCode.Eq:
      row = simple('NIL');
      break;
    case OpCode.Gt:
      row = simple('GT');
      break;
    case OpCode.Ge:
      row = simple('GE');
      break;
    case OpCode.Lt:
      row = simple('LT');
      break;
    case OpCode.Le:
      row = simple('LE');
      break;
    case OpCode.Ne:
      row = simple('NE');
      break;
    case OpCode.Not:
      row = simple('NOT');
      break;
    case OpCode.Add:
      row = simple('ADD');
      break;
    case OpCode.Sub:
      row = simple('SUB');
      break;
    case OpCode.Mul:
      row = simple('MUL');
      break;
    case OpCode.Div:
      row = simple('DIV');
      break;
    case OpCode.Neg:
      row = simple('NEG');
      break;
    case OpCode.Print:
      row = simple('PRINT');
      break;
    case OpCode.Exit:
      row = simple('EXIT');
      break;
    case OpCode.Jump:
      row = jump('JUMP', FORWARD);
      break;
    case OpCode.JumpIfFalse:
      row = jump('JUMP_IF_FALSE', FORWARD);
      break;
    case OpCode.Loop:
      row = simple('LOOP');
      break;
    case OpCode.Return:
      row = simple('RETURN');
      break;
    default:
      row = [String(lineNo), String(offset), 'Unknown opcode', toHex(code)];
  }

  return [offset, row];
}
