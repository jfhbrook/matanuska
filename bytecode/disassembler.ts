import table from 'text-table';

import { formatter } from '../format';
import { OpCode } from './opcodes';
import { Chunk } from './chunk';

type LineNo = string;
type Offset = string;
type Code = string;
type Row = [LineNo, Offset, Code, string];

function chunkName(chunk: Chunk): string {
  return chunk.filename;
}

function toHex(value: number): string {
  return `0x${value.toString(16)}`;
}

export function disassemble(chunk: Chunk): string {
  const data: Row[] = [];
  const header = `=== Disassembly of ${chunkName(chunk)}: ===`;

  let offset = 0;
  let row: Row;

  while (offset < chunk.code.length) {
    [offset, row] = _disassembleInstruction(chunk, offset);
    data.push(row);
  }

  return `${header}\n${table(data)}`;
}

export function disassembleInstruction(chunk: Chunk, offset: number): string {
  const row = _disassembleInstruction(chunk, offset)[1];
  return table([row]);
}

function _disassembleInstruction(chunk: Chunk, offset: number): [number, Row] {
  let lineNo = 0;
  let code: number = 0xff;
  let row: Row;

  function advance(): number {
    offset++;
    code = chunk.code[offset - 1];
    lineNo = chunk.lines[offset - 1];
    return code;
  }

  function simple(code: string): Row {
    return [String(lineNo), String(offset), code, ''];
  }

  function constant(): Row {
    return [
      String(lineNo),
      String(offset),
      'CONSTANT',
      formatter.format(chunk.constants[advance()]),
    ];
  }

  switch (advance()) {
    case OpCode.Constant:
      row = constant();
      break;
    case OpCode.Nil:
      row = simple('NIL');
      break;
    case OpCode.True:
      row = simple('TRUE');
      break;
    case OpCode.False:
      row = simple('FALSE');
      break;
    case OpCode.Pop:
      row = simple('POP');
      break;
    case OpCode.Eq:
      row = simple('NIL');
      break;
    case OpCode.Gt:
      row = simple('GT');
      break;
    case OpCode.Ge:
      row = simple('GE');
      break;
    case OpCode.Lt:
      row = simple('LT');
      break;
    case OpCode.Le:
      row = simple('LE');
      break;
    case OpCode.Ne:
      row = simple('NE');
      break;
    case OpCode.Not:
      row = simple('NOT');
      break;
    case OpCode.Add:
      row = simple('ADD');
      break;
    case OpCode.Sub:
      row = simple('SUB');
      break;
    case OpCode.Mul:
      row = simple('MUL');
      break;
    case OpCode.Div:
      row = simple('DIV');
      break;
    case OpCode.Neg:
      row = simple('NEG');
      break;
    case OpCode.Print:
      row = simple('PRINT');
      break;
    case OpCode.Exit:
      row = simple('EXIT');
      break;
    case OpCode.Jump:
      row = simple('JUMP');
      break;
    case OpCode.JumpIfFalse:
      row = simple('JUMP_IF_FALSE');
      break;
    case OpCode.Loop:
      row = simple('LOOP');
      break;
    case OpCode.Return:
      row = simple('RETURN');
      break;
    default:
      row = [String(lineNo), String(offset), 'Unknown opcode', toHex(code)];
  }

  return [offset, row];
}
