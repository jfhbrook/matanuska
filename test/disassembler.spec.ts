import { describe, expect, test } from 'vitest';

import { shortToBytes } from '../bytecode/short';
import { OpCode } from '../bytecode/opcodes';
import { disassemble } from '../bytecode/disassembler';

import { chunk } from './helpers/bytecode';

describe('expressions', () => {
  test('255', () => {
    expect(
      disassemble(
        chunk({
          constants: [255],
          code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('123.456', () => {
    expect(
      disassemble(
        chunk({
          constants: [123.456],
          code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  for (const bool of [true, false]) {
    test(`${bool}`, () => {
      expect(
        disassemble(
          chunk({
            constants: [bool],
            code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
            lines: [-1, -1, -1, -1, -1],
          }),
        ),
      ).toMatchSnapshot();
    });
  }

  test('nil', () => {
    expect(
      disassemble(
        chunk({
          constants: [],
          code: [OpCode.Nil, OpCode.Pop, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('"hello world"', () => {
    expect(
      disassemble(
        chunk({
          constants: ['hello world'],
          code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('(1)', () => {
    expect(
      disassemble(
        chunk({
          constants: [1],
          code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('1 + 1', () => {
    expect(
      disassemble(
        chunk({
          constants: [1, 1],
          code: [
            OpCode.Constant,
            0,
            OpCode.Constant,
            1,
            OpCode.Add,
            OpCode.Pop,
            OpCode.Nil,
            OpCode.Return,
          ],
          lines: [-1, -1, -1, -1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('-1', () => {
    expect(
      disassemble(
        chunk({
          constants: [1],
          code: [
            OpCode.Constant,
            0,
            OpCode.Neg,
            OpCode.Pop,
            OpCode.Nil,
            OpCode.Return,
          ],
          lines: [-1, -1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('1 : 1', () => {
    expect(
      disassemble(
        chunk({
          constants: [1, 1],
          code: [
            OpCode.Constant,
            0,
            OpCode.Pop,
            OpCode.Constant,
            1,
            OpCode.Pop,
            OpCode.Nil,
            OpCode.Return,
          ],
          lines: [-1, -1, -1, -1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });
});

describe('print', () => {
  test('print 255', () => {
    expect(
      disassemble(
        chunk({
          constants: [255],
          code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('print 123.456', () => {
    expect(
      disassemble(
        chunk({
          constants: [123.456],
          code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  for (const bool of [true, false]) {
    test(`print ${bool}`, () => {
      expect(
        disassemble(
          chunk({
            constants: [bool],
            code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
            lines: [-1, -1, -1, -1, -1],
          }),
        ),
      ).toMatchSnapshot();
    });
  }

  test('print nil', () => {
    expect(
      disassemble(
        chunk({
          constants: [],
          code: [OpCode.Nil, OpCode.Print, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('print "hello world"', () => {
    expect(
      disassemble(
        chunk({
          constants: ['hello world'],
          code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('print (1)', () => {
    expect(
      disassemble(
        chunk({
          constants: [1],
          code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('print 1 + 1', () => {
    expect(
      disassemble(
        chunk({
          constants: [1, 1],
          code: [
            OpCode.Constant,
            0,
            OpCode.Constant,
            1,
            OpCode.Add,
            OpCode.Print,
            OpCode.Nil,
            OpCode.Return,
          ],
          lines: [-1, -1, -1, -1, -1, -1, -1, -1],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('print i% + 1', () => {
    expect(
      disassemble(
        chunk({
          constants: ['i%', 1],
          code: [
            OpCode.Constant,
            0,
            OpCode.GetGlobal,
            0,
            OpCode.Constant,
            1,
            OpCode.Add,
          ],
          lines: [100, 100, 100, 100, 100, 100, 100],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('let i% = 1', () => {
    expect(
      disassemble(
        chunk({
          constants: ['i%', 1],
          code: [
            OpCode.Constant,
            0,
            OpCode.Constant,
            1,
            OpCode.DefineGlobal,
            0,
            OpCode.Nil,
            OpCode.Return,
          ],
          lines: [100, 100, 100, 100, 100, 100, 100, 100],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('i% = 1', () => {
    expect(
      disassemble(
        chunk({
          constants: ['i%', 1],
          code: [
            OpCode.Constant,
            0,
            OpCode.Constant,
            1,
            OpCode.SetGlobal,
            0,
            OpCode.Nil,
            OpCode.Return,
          ],
          lines: [100, 100, 100, 100, 100, 100, 100, 100],
        }),
      ),
    ).toMatchSnapshot();
  });

  test('if true then print "true" else print "false" endif', () => {
    expect(
      disassemble(
        chunk({
          constants: [true, 'true', 'false'],
          code: [
            OpCode.Constant,
            0,
            // Jump to "else"
            OpCode.JumpIfFalse,
            ...shortToBytes(7),
            // "then" block
            OpCode.Pop,
            OpCode.Constant,
            1,
            OpCode.Print,
            // Jump to end
            OpCode.Jump,
            ...shortToBytes(4),
            // "else" block
            OpCode.Pop,
            OpCode.Constant,
            2,
            OpCode.Print,
            OpCode.Nil,
            OpCode.Return,
          ],
          lines: [
            100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100, 100,
            100, 100, 100, 100, 100,
          ],
        }),
      ),
    ).toMatchSnapshot();
  });
});

test('simple program', () => {
  expect(
    disassemble(
      chunk({
        constants: ['hello world', 'goodbye'],
        code: [
          OpCode.Constant,
          0,
          OpCode.Print,
          OpCode.Constant,
          1,
          OpCode.Print,
          OpCode.Nil,
          OpCode.Return,
        ],
        lines: [100, 100, 100, 200, 200, 200, 200, 200],
      }),
    ),
  ).toMatchSnapshot();
});

test('malformed lines', () => {
  expect(
    disassemble(
      chunk({
        constants: ['hello world'],
        code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
        lines: [],
      }),
    ),
  ).toMatchSnapshot();
});

test('missing constants', () => {
  expect(
    disassemble(
      chunk({
        constants: [],
        code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    ),
  ).toMatchSnapshot();
});

test('unknown opcode', () => {
  expect(
    disassemble(
      chunk({
        constants: [],
        // It's over nine thousand!!!
        code: [9001 as OpCode],
        lines: [-1],
      }),
    ),
  ).toMatchSnapshot();
});
