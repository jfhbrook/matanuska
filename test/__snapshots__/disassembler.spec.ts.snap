// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expressions > "hello world" 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  hello world
-1  3  POP
-1  4  NIL
-1  5  RETURN"
`;

exports[`expressions > (1) 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  1
-1  3  POP
-1  4  NIL
-1  5  RETURN"
`;

exports[`expressions > -1 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  1
-1  3  NEG
-1  4  POP
-1  5  NIL
-1  6  RETURN"
`;

exports[`expressions > 1 + 1 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  1
-1  3  CONSTANT  1
-1  5  ADD
-1  6  POP
-1  7  NIL
-1  8  RETURN"
`;

exports[`expressions > 1 : 1 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  1
-1  3  POP
-1  4  CONSTANT  1
-1  6  POP
-1  7  NIL
-1  8  RETURN"
`;

exports[`expressions > 123.456 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  123.456
-1  3  POP
-1  4  NIL
-1  5  RETURN"
`;

exports[`expressions > 255 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  255
-1  3  POP
-1  4  NIL
-1  5  RETURN"
`;

exports[`expressions > false 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  false
-1  3  POP
-1  4  NIL
-1  5  RETURN"
`;

exports[`expressions > nil 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  NIL
-1  2  POP
-1  3  NIL
-1  4  RETURN"
`;

exports[`expressions > true 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  true
-1  3  POP
-1  4  NIL
-1  5  RETURN"
`;

exports[`malformed lines 1`] = `
"=== Disassembly of <unknown>: ===
undefined  1  CONSTANT  hello world
undefined  3  POP
undefined  4  NIL
undefined  5  RETURN"
`;

exports[`missing constants 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  undefined
-1  3  POP
-1  4  NIL
-1  5  RETURN"
`;

exports[`print > i% = 1 1`] = `
"=== Disassembly of <unknown>: ===
100  1  CONSTANT    i%
100  3  CONSTANT    1
100  5  SET_GLOBAL  i%
100  7  NIL
100  8  RETURN"
`;

exports[`print > if true then print "true" else print "false" endif 1`] = `
"=== Disassembly of <unknown>: ===
100  1   CONSTANT       true
100  3   JUMP_IF_FALSE  3 -> 13
100  6   POP
100  7   CONSTANT       true
100  9   PRINT
100  10  JUMP           10 -> 17
100  13  POP
100  14  CONSTANT       false
100  16  PRINT
100  17  NIL
100  18  RETURN"
`;

exports[`print > let i% = 1 1`] = `
"=== Disassembly of <unknown>: ===
100  1  CONSTANT       i%
100  3  CONSTANT       1
100  5  DEFINE_GLOBAL  i%
100  7  NIL
100  8  RETURN"
`;

exports[`print > print "hello world" 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  hello world
-1  3  PRINT
-1  4  NIL
-1  5  RETURN"
`;

exports[`print > print (1) 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  1
-1  3  PRINT
-1  4  NIL
-1  5  RETURN"
`;

exports[`print > print 1 + 1 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  1
-1  3  CONSTANT  1
-1  5  ADD
-1  6  PRINT
-1  7  NIL
-1  8  RETURN"
`;

exports[`print > print 123.456 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  123.456
-1  3  PRINT
-1  4  NIL
-1  5  RETURN"
`;

exports[`print > print 255 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  255
-1  3  PRINT
-1  4  NIL
-1  5  RETURN"
`;

exports[`print > print false 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  false
-1  3  PRINT
-1  4  NIL
-1  5  RETURN"
`;

exports[`print > print i% + 1 1`] = `
"=== Disassembly of <unknown>: ===
100  1  CONSTANT    i%
100  3  GET_GLOBAL  i%
100  5  CONSTANT    1
100  7  ADD"
`;

exports[`print > print nil 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  NIL
-1  2  PRINT
-1  3  NIL
-1  4  RETURN"
`;

exports[`print > print true 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  CONSTANT  true
-1  3  PRINT
-1  4  NIL
-1  5  RETURN"
`;

exports[`simple program 1`] = `
"=== Disassembly of <unknown>: ===
100  1  CONSTANT  hello world
100  3  PRINT
200  4  CONSTANT  goodbye
200  6  PRINT
200  7  NIL
200  8  RETURN"
`;

exports[`unknown opcode 1`] = `
"=== Disassembly of <unknown>: ===
-1  1  Unknown opcode  0x2329"
`;
